### ISSUE TITLE
Compute error for `PSNR` and `SSIM`

### ISSUE THREAD

**Issue by leftthomas** [photosynthesis-team/piq#288]
Convert RGB image to YCbCr and take luminance should be `Y = 0.256789 R + 0.504129 G +0.097906 B`, not
https://github.com/photosynthesis-team/piq/blob/e8f6513afa8f9138bcd19c0928763cc18bf7e060/piq/psnr.py#L39

As for `SSIM`, to make sure the results can be same with [Matlab Code](https://github.com/hongwang01/RCDNet/tree/master/Performance_evaluation), it must use `DoubleTensor`, and these three lines should change:
https://github.com/photosynthesis-team/piq/blob/e8f6513afa8f9138bcd19c0928763cc18bf7e060/piq/ssim.py#L55
https://github.com/photosynthesis-team/piq/blob/e8f6513afa8f9138bcd19c0928763cc18bf7e060/piq/ssim.py#L56
https://github.com/photosynthesis-team/piq/blob/e8f6513afa8f9138bcd19c0928763cc18bf7e060/piq/functional/filters.py#L35

Please refer to [my implement](https://github.com/leftthomas/RCDNet/blob/master/utils.py) of `SSIM` and `PSNR`.

**Comment by zakajd** [comment]
Hi @leftthomas thanks for raising the issue.
I can't yet agree with your point on PSNR conversion formula. Can you please provide a reference that mentions `Y = 0.256789 R + 0.504129 G +0.097906 B` values? Wikipedia pages on both [YUV](https://en.wikipedia.org/wiki/YUV) and [YCbCr](https://en.wikipedia.org/wiki/YCbCr) mention values we currently use with reference to ITU-R BT.601 standard. 

With regard to DoubleTensor — how much is the absolute difference between 2 methods using single and double precisions? 
BTW link to your implementation shows 404 error.

**Comment by leftthomas** [comment]
@zakajd I have make the repo of my implementation public, and you could refer to. 

With regard to DoubleTensor — I have tested with `Rain1400` in terms of `SSIM`, which is a Derain dataset, the absolute difference between 2 methods using single and double precisions is 0.0003, this value can not be ignored for `Derain task`, and I think you could refactor the API as like this:
``
coords = torch.arange(kernel_size, dtype=input.dtype, device=input.device)
``
this could compatible with both `FloatTensor` and `DoubleTensor`, according to input data.

About Y channel compute for `YCbCr`, I recommend you to refer the part of ` ITU-R BT.601` and [Matlab Doc](https://www.mathworks.com/help/images/ref/rgb2ycbcr.html) or [this link](https://blog.csdn.net/linda_dinglei/article/details/109988844)
<img width="890" alt="image" src="https://user-images.githubusercontent.com/9991443/152270018-33e92e5c-06e6-429e-a977-6cec1126f0e2.png">

**Comment by leftthomas** [comment]
@zakajd Can you tell me the bug is fixed or not?  And about the compute of `YCbCr`.

**Comment by zakajd** [comment]
@leftthomas PR #292 adds dtype propagation though computation. You can fork it and use in your project or wait until it's merged into master (this will take up to 1 week). 

Computation of Y channel in PSNR won't be changed due to insufficient proofs. Neither PIL ([docs](https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.Image.convert)), neither OpenCV ([docs](https://docs.opencv.org/2.4/modules/imgproc/doc/miscellaneous_transformations.html#void%20cvtColor%28InputArray%20src,%20OutputArray%20dst,%20int%20code,%20int%20dstCn%29)) use the formula you provided and I assume those guys know what they are doing.

**Comment by leftthomas** [comment]
@zakajd about the `Y` channel compute, Python code is not same with Matlab, and some tasks use the Matlab code to obtain the metrics, and I think maybe you could add an args to let user to choose which implement to use, such as `convert_to_y='python'` or `convert_to_y='matlab'` , I suggest you to refer the implement of Matlab:
https://github.com/victorca25/traiNNer/blob/12d006fd44ed304e4178839c53b1f3d95ca25dcb/metrics/calculate_PSNR_SSIM.py#L122
and 
https://github.com/victorca25/traiNNer/blob/master/metrics/calculate_PSNR_SSIM.m

**Comment by leftthomas** [comment]
@zakajd , This is another implement about the `Y` channel compute
https://github.com/chaofengc/IQA-PyTorch/blob/1f1ba21e73f6039bacdcd6e72147193ac9cae4ef/pyiqa/utils/color_util.py#L49

**Comment by denproc** [comment]
@leftthomas thank you for your feedback regarding the PSNR and SSIM performance. 
PSNR compares grayscale images. In order to get the grayscale image, we use conversion proposed in PIL ([docs](https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.Image.convert)) and OpenCV ([docs](https://docs.opencv.org/2.4/modules/imgproc/doc/miscellaneous_transformations.html#void%20cvtColor%28InputArray%20src,%20OutputArray%20dst,%20int%20code,%20int%20dstCn%29)). This conversion is identical to Y channel of RGB2YIQ transformation. Sadly, previous comment in code was misleading and will be fixed to avoid misunderstanding in future.

Following your quote, [PSNR](https://github.com/chaofengc/IQA-PyTorch/blob/1f1ba21e73f6039bacdcd6e72147193ac9cae4ef/pyiqa/archs/psnr_arch.py#L20) implementation proposed in IQA-PyTorch uses RBG2YIQ as a default colour conversion, which matches our approach. 
Any additional colour conversions other than to standard grayscale could be added as preprocessing step to your code.

**Comment by leftthomas** [comment]
@denproc, the value used in [PSNR](https://github.com/chaofengc/IQA-PyTorch/blob/1f1ba21e73f6039bacdcd6e72147193ac9cae4ef/pyiqa/utils/color_util.py#L49) implementation proposed in IQA-PyTorch is same as `Matlab: Y = 0.256789 R + 0.504129 G +0.097906 B` (after divided by 255), but in this repo is `rgb_to_grey = torch.tensor([0.299, 0.587, 0.114]).view(1, -1, 1, 1).to(x)`

**Comment by zakajd** [comment]
> PSNR should be computed on Y channel of YCbCr, not YIQ

Any source for such claim? 
This metric is ~30 years old and there is no single reference implementation that should be matched. For RGB images it can be computed on all channels independently and then averaged, or use Y channel of YIQ, or Y channel of YCbCr, or Hue channel of HSL. All this is just a _design choice_ that authors make. 
**We made ours to compute on Y channel of YIQ.**

You can fork and modify the script to suit your specific needs.

**Comment by leftthomas** [comment]
@zakajd I mean if matching the Matlab version of `SSIM` and `PSNR` based on `Y` channel, then the compute of `Y` should be like what I have given, just like this
https://github.com/victorca25/traiNNer/blob/master/metrics/calculate_PSNR_SSIM.m

**Comment by zakajd** [comment]
@leftthomas this doesn't prove anything.
[Official MATLAB documentation](https://www.mathworks.com/help/images/ref/psnr.html) says nothing about handling of RGB images and what conversion it uses.
Code you shared is just another project, whose authors made a _choice_ of computing in YCbCr space.

**Comment by snk4tr** [comment]
> I mean if matching the Matlab version of SSIM and PSNR

Dear @leftthomas, the Matplab version that you reference is not the official one. It is a **custom** implementation where the authors made **their choice** to use the `RGB2YCbCr` conversion. However, what we refer to are the **official** Matlab implementation [here](https://www.mathworks.com/help/images/ref/psnr.html), the **official** PIL implementation [here](https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.Image.convert) and the **official** OpenCV implementation [here](https://docs.opencv.org/2.4/modules/imgproc/doc/miscellaneous_transformations.html#void%20cvtColor%28InputArray%20src,%20OutputArray%20dst,%20int%20code,%20int%20dstCn%29).

Since those implementations are the ones that the whole research community refer to, we doubt that the implementation you propose is valuable enough to be included. To me it seems like yet another way to implement the same metric. If you are not agree, please provide sufficient references.

**Comment by leftthomas** [comment]
@zakajd @snk4tr I have tried the `rgb2ycbcr` of Matlab and the same function of OpenCV or PIL, and you will find the results is not same.

**Comment by denproc** [comment]
@leftthomas we are discussing two different formats. You chose to use 'YCbCr' colour map.

In our project, we _chose_ luminance channel, which is `Y` channel of `YIQ` colour map. [MATLAB YIQ colour map (rgb2ntsc)](https://www.mathworks.com/help/images/ref/rgb2ntsc.html), [MATLAB rgb2gray](https://www.mathworks.com/help/matlab/ref/rgb2gray.html?s_tid=doc_ta), [OpenCV](https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.Image.convert) and [PIL](https://docs.opencv.org/2.4/modules/imgproc/doc/miscellaneous_transformations.html#void%20cvtColor%28InputArray%20src,%20OutputArray%20dst,%20int%20code,%20int%20dstCn%29) are using the same weights to get luminance channel. 

I have tried the `rgb2ntsc` of Matlab and the same function of OpenCV or PIL, and you will find the results are the same.

**Comment by leftthomas** [comment]
@denproc ok, but I mean the implement of `rgb2ycbcr` is not same in Matlab and OpenCV or PIL

**Comment by denproc** [comment]
@leftthomas, Indeed, `RGB2YCrCb` uses different from `RGB2YIQ` (or `rgb2ntsc`) weights. However, we are not using `rgb2ycbcr` colour conversion. Our approach uses more common `YIQ` colour map.
